File: Source/Client/Modules/Objects/Resource.vb
Imports System.Drawing
Imports System.IO
Imports Mirage.Sharp.Asfw
Imports Core

Module Resource

#Region "Globals & Type"

    Friend ResourceIndex As Integer
    Friend ResourcesInit As Boolean

#End Region

#Region "Database"

    Sub ClearResource(index As Integer)
        Type.Resource(index) = Nothing
        Type.Resource(index).Name = ""
        Resource_Loaded(index) = False
    End Sub

    Sub ClearResources()
        Dim i As Integer

       For i = 1 To MAX_RESOURCES
            ClearResource(i)
        Next

    End Sub

    Sub StreamResource(resourceNum As Integer)
        If resourceNum > 0 And Type.Resource(resourceNum).Name = "" Or Resource_Loaded(resourceNum) = False Then
            Resource_Loaded(resourceNum) = True
            SendRequestResource(resourceNum)
        End If
    End Sub

#End Region

#Region "Incoming Packets"

    Sub Packet_MapResource(ByRef data() As Byte)
        Dim i As Integer
        Dim buffer As New ByteStream(data)
        ResourceIndex = buffer.ReadInt32
        ResourcesInit = False

        If ResourceIndex > 0 Then
            ReDim Preserve MapResource(ResourceIndex)

            For i = 0 To ResourceIndex
                MyMapResource(i).State = buffer.ReadByte
                MyMapResource(i).X = buffer.ReadInt32
                MyMapResource(i).Y = buffer.ReadInt32
            Next

            ResourcesInit = True
        End If

        buffer.Dispose()
    End Sub

    Sub Packet_UpdateResource(ByRef data() As Byte)
        Dim resourceNum As Integer
        Dim buffer As New ByteStream(data)
        resourceNum = buffer.ReadInt32

        Type.Resource(resourceNum).Animation = buffer.ReadInt32()
        Type.Resource(resourceNum).EmptyMessage = buffer.ReadString()
        Type.Resource(resourceNum).ExhaustedImage = buffer.ReadInt32()
        Type.Resource(resourceNum).Health = buffer.ReadInt32()
        Type.Resource(resourceNum).ExpReward = buffer.ReadInt32()
        Type.Resource(resourceNum).ItemReward = buffer.ReadInt32()
        Type.Resource(resourceNum).Name = buffer.ReadString()
        Type.Resource(resourceNum).ResourceImage = buffer.ReadInt32()
        Type.Resource(resourceNum).ResourceType = buffer.ReadInt32()
        Type.Resource(resourceNum).RespawnTime = buffer.ReadInt32()
        Type.Resource(resourceNum).SuccessMessage = buffer.ReadString()
        Type.Resource(resourceNum).LvlRequired = buffer.ReadInt32()
        Type.Resource(resourceNum).ToolRequired = buffer.ReadInt32()
        Type.Resource(resourceNum).Walkthrough = buffer.ReadInt32()

        buffer.Dispose()
    End Sub

#End Region

#Region "Outgoing Packets"

    Sub SendRequestResource(resourceNum As Integer)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CRequestResource)

        buffer.WriteInt32(resourcenum)
        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

#End Region

#Region "Drawing"

    Friend Sub DrawResource(resource As Integer, dx As Integer, dy As Integer, rec As Rectangle)
        Dim x As Integer
        Dim y As Integer
        Dim width As Integer
        Dim height As Integer

        If resource < 1 Or resource > NumResources Then Exit Sub

        x = ConvertMapX(dx)
        y = ConvertMapY(dy)
        width = (rec.Right - rec.Left)
        height = (rec.Bottom - rec.Top)

        If rec.Width < 0 Or rec.Height < 0 Then Exit Sub

        Client.EnqueueTexture(System.IO.Path.Combine(Core.Path.Resources, resource), x, y, rec.X, rec.Y, rec.Width, rec.Height, rec.Width, rec.Height)
    End Sub

    Friend Sub DrawMapResource(resourceNum As Integer)
        Dim resourceMaster As Integer

        Dim resourceState As Integer
        Dim resourceSprite As Integer
        Dim rec As Rectangle
        Dim x As Integer, y As Integer

        If GettingMap Then Exit Sub
       If MapData = False Then Exit Sub

        If MyMapResource(resourceNum).X > MyMap.MaxX Or MyMapResource(resourceNum).Y > MyMap.MaxY Then Exit Sub

        ' Get the Resource type
        resourceMaster = MyMap.Tile(MyMapResource(resourceNum).X, MyMapResource(resourceNum).Y).Data1

        If resourceMaster = 0 Then Exit Sub

        If Type.Resource(resourceMaster).ResourceImage = 0 Then Exit Sub

        StreamResource(resourceMaster)

        ' Get the Resource state
        resourceState = MyMapResource(resourceNum).State

        If resourceState = 0 Then ' normal
            resourceSprite = Type.Resource(resourceMaster).ResourceImage
        ElseIf resourceState = 1 Then ' used
            resourceSprite = Type.Resource(resourceMaster).ExhaustedImage
        End If

        ' src rect
        With rec
            .Y = 0
            .Height = Client.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite)).Height
            .X = 0
            .Width = Client.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite)).Width
        End With

        ' Set base x + y, then the offset due to size
        x = (MyMapResource(resourceNum).X * PicX) - (Client.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite)).Width / 2) + 16
        y = (MyMapResource(resourceNum).Y * PicY) - Client.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite)).Height + 32

        DrawResource(resourceSprite, x, y, rec)
    End Sub

#End Region

End Module

File: Source/Client/Game/Objects/Resource.cs
using System;
using System.Drawing;
using Core;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;

namespace Client
{

    public class MapResource
    {

        #region Database

        public static void ClearResource(int index)
        {
            Data.Resource[index] = default;
            Data.Resource[index].Name = "";
            GameState.Resource_Loaded[index] = 0;
        }

        public static void ClearResources()
        {
            Array.Resize(ref Data.Resource, Constant.MAX_RESOURCES);

            for (int i = 0; i < Constant.MAX_RESOURCES; i++)
                ClearResource(i);

        }

        public static void StreamResource(int resourceNum)
        {
            if (resourceNum >= 0 && string.IsNullOrEmpty(Data.Resource[resourceNum].Name) && GameState.Resource_Loaded[resourceNum] == 0)
            {
                GameState.Resource_Loaded[resourceNum] = 1;
                SendRequestResource(resourceNum);
            }
        }

        #endregion

        #region Incoming Packets

        public static void Packet_MapResource(ref byte[] data)
        {
            int i;
            var buffer = new ByteStream(data);
            GameState.ResourceIndex = buffer.ReadInt32();
            GameState.ResourcesInit = false;

            if (GameState.ResourceIndex > 0)
            {
                Array.Resize(ref Data.MapResource, GameState.ResourceIndex);
                Array.Resize(ref Data.MyMapResource, GameState.ResourceIndex);

                var loopTo = GameState.ResourceIndex;
                for (i = 0; i < loopTo; i++)
                {
                    Data.MyMapResource[i].State = buffer.ReadByte();
                    Data.MyMapResource[i].X = buffer.ReadInt32();
                    Data.MyMapResource[i].Y = buffer.ReadInt32();
                }

                GameState.ResourcesInit = true;
            }

            buffer.Dispose();
        }

        public static void Packet_UpdateResource(ref byte[] data)
        {
            int resourceNum;
            var buffer = new ByteStream(data);
            resourceNum = buffer.ReadInt32();

            Data.Resource[resourceNum].Animation = buffer.ReadInt32();
            Data.Resource[resourceNum].EmptyMessage = buffer.ReadString();
            Data.Resource[resourceNum].ExhaustedImage = buffer.ReadInt32();
            Data.Resource[resourceNum].Health = buffer.ReadInt32();
            Data.Resource[resourceNum].ExpReward = buffer.ReadInt32();
            Data.Resource[resourceNum].ItemReward = buffer.ReadInt32();
            Data.Resource[resourceNum].Name = buffer.ReadString();
            Data.Resource[resourceNum].ResourceImage = buffer.ReadInt32();
            Data.Resource[resourceNum].ResourceType = buffer.ReadInt32();
            Data.Resource[resourceNum].RespawnTime = buffer.ReadInt32();
            Data.Resource[resourceNum].SuccessMessage = buffer.ReadString();
            Data.Resource[resourceNum].LvlRequired = buffer.ReadInt32();
            Data.Resource[resourceNum].ToolRequired = buffer.ReadInt32();
            Data.Resource[resourceNum].Walkthrough = Conversions.ToBoolean(buffer.ReadInt32());

            buffer.Dispose();
        }

        #endregion

        #region Outgoing Packets

        public static void SendRequestResource(int resourceNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CRequestResource);

            buffer.WriteInt32(resourceNum);
            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        #endregion

        #region Drawing

        public static void DrawResource(int resource, int dx, int dy, Rectangle rec)
        {
            int x;
            int y;
            int width;
            int height;

            if (resource < 1 | resource > GameState.NumResources)
                return;

            x = GameLogic.ConvertMapX(dx);
            y = GameLogic.ConvertMapY(dy);
            width = rec.Right - rec.Left;
            height = rec.Bottom - rec.Top;

            if (rec.Width < 0 | rec.Height < 0)
                return;

            string argpath = System.IO.Path.Combine(Core.Path.Resources, resource.ToString());
            GameClient.RenderTexture(ref argpath, x, y, rec.X, rec.Y, rec.Width, rec.Height, rec.Width, rec.Height);
        }

        public static void DrawMapResource(int resourceNum)
        {
            int mapResourceNum;
            int resourceState;
            var resourceSprite = default(int);
            var rec = default(Rectangle);
            int x;
            int y;

            if (GameState.GettingMap)
                return;

            if (!GameState.MapData)
                return;

            if (Data.MyMapResource[resourceNum].X > Data.MyMap.MaxX | Data.MyMapResource[resourceNum].Y > Data.MyMap.MaxY)
                return;

            mapResourceNum = Data.MyMap.Tile[Data.MyMapResource[resourceNum].X, Data.MyMapResource[resourceNum].Y].Data1;

            if (mapResourceNum == 0)
                mapResourceNum = Data.MyMap.Tile[Data.MyMapResource[resourceNum].X, Data.MyMapResource[resourceNum].Y].Data1_2;

            StreamResource(mapResourceNum);

            if (Data.Resource[mapResourceNum].ResourceImage == 0)
                return;

            // Get the Resource state
            resourceState = Data.MyMapResource[resourceNum].State;

            if (resourceState == 0) // normal
            {
                resourceSprite = Data.Resource[mapResourceNum].ResourceImage;
            }
            else if (resourceState == 1) // used
            {
                resourceSprite = Data.Resource[mapResourceNum].ExhaustedImage;
            }

            // src rect
            rec.Y = 0;
            rec.Height = GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite.ToString())).Height;
            rec.X = 0;
            rec.Width = GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite.ToString())).Width;

            // Set base x + y, then the offset due to size
            x = (int)Math.Round(Data.MyMapResource[resourceNum].X * GameState.PicX - GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite.ToString())).Width / 2d + 16d);
            y = Data.MyMapResource[resourceNum].Y * GameState.PicY - GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Resources, resourceSprite.ToString())).Height + 32;

            DrawResource(resourceSprite, x, y, rec);
        }

        #endregion

    }
}